saga-spring/src/main/java/io/servicecomb/saga/spring/SagaController.java

```diff
@@ -125,36 +148,90 @@ public SagaController(SagaExecutionComponent sagaExecutionComponent, SagaEventRe
-    Iterable<SagaEventEntity> entities = repo.findBySagaId(sagaId);
-    List<SagaEventVo> events = new ArrayList<>();
-    entities
-        .forEach(e -> events.add(new SagaEventVo(e.id(), e.sagaId(), e.creationTime(), e.type(), e.contentJson())));
-    return ResponseEntity.ok(events);
+  public ResponseEntity<SagaExecutionDetail> queryExecutionDetail(@PathVariable String sagaId) {
+    SagaEventEntity[] entities = repo.findBySagaId(sagaId).toArray(new SagaEventEntity[0]);
+    Optional<SagaEventEntity> sagaStartEvent = Arrays.stream(entities)
+        .filter(entity -> SagaStartedEvent.class.getSimpleName().equals(entity.type())).findFirst();
+    Map<String, List<String>> router = new HashMap<>();
+    Map<String, String> status = new HashMap<>();
+    Map<String, String> error = new HashMap<>();
+    if (sagaStartEvent.isPresent()) {
+      SagaDefinition definition = fromJsonFormat.fromJson(sagaStartEvent.get().contentJson());
+      SingleLeafDirectedAcyclicGraph<SagaRequest> graph = graphBuilder
+          .build(definition.requests());
+      loopLoadGraphNodes(router, graph.root());
+
+      Collection<SagaEventEntity> transactionAbortEvents = Arrays.stream(entities)
+          .filter(entity -> TransactionAbortedEvent.class.getSimpleName().equals(entity.type())).collect(
+              Collectors.toList());
+      for (SagaEventEntity transactionAbortEvent : transactionAbortEvents) {
+        try {
+          JsonNode root = mapper.readTree(transactionAbortEvent.contentJson());
+          String id = root.at("/request/id").asText();
+          status.put(id, "Failed");
+          error.put(id, root.at("/response/body").asText());
+        } catch (IOException ignored) {
+        }
+      }
+
+      Collection<SagaEventEntity> transactionEndEvents = Arrays.stream(entities)
+          .filter(entity -> TransactionEndedEvent.class.getSimpleName().equals(entity.type())).collect(
+              Collectors.toList());
+      for (SagaEventEntity transactionEndEvent : transactionEndEvents) {
+        try {
+          JsonNode root = mapper.readTree(transactionEndEvent.contentJson());
+          status.put(root.at("/request/id").asText(), "OK");
+        } catch (IOException ignored) {
+        }
+      }
+    }
+    return ResponseEntity.ok(new SagaExecutionDetail(router, status, error));
+  }
+
+  private void loopLoadGraphNodes(Map<String, List<String>> router,
+      Node<SagaRequest> node) {
+    if (node != null) {
+      if (isNodeValid(node)) {
+        List<String> point = router.computeIfAbsent(node.value().id(), key -> new ArrayList<>());
+        for (Node<SagaRequest> child : node.children()) {
+          point.add(child.value().id());
+          loopLoadGraphNodes(router, child);
+        }
+      }
+    }
+  }
+
+  private boolean isNodeValid(Node<SagaRequest> node) {
+    return !NoOpSagaRequest.SAGA_END_REQUEST.id().equals(node.value().id())
+        && node.children() != null && node.children().size() != 0;
```

Comment:
only saga end request can have empty children. empty and null checking are duplicate with end request id checking

Reply:
Done